iterable-extension/src/test/java/com/mparticle/ext/iterable/IterableExtensionTest.java

```diff
@@ -448,4 +446,106 @@ public void testProcessProductActionEvent() throws Exception {
         assertEquals(trackPurchaseRequest.items.size(), 2);
         assertEquals(trackPurchaseRequest.total, new BigDecimal(101d));
     }
+
+    @Test
+    public void testGetPlaceholderEmailNoEnvironmentOrStamp() throws Exception {
+        EventProcessingRequest request = new EventProcessingRequest();
+        request.setRuntimeEnvironment(null);
+        request.setDeviceApplicationStamp(null);
+        Exception e = null;
+        try {
+            String email = IterableExtension.getPlaceholderEmail(request);
+        }catch (IOException ioe) {
+            e = ioe;
+        }
+        assertNotNull(e);
+    }
+
+    @Test
+    public void testGetPlaceholderEmailNoEnvironment() throws Exception {
+        EventProcessingRequest request = new EventProcessingRequest();
+        request.setRuntimeEnvironment(null);
+        request.setDeviceApplicationStamp("1234");
+        String email = IterableExtension.getPlaceholderEmail(request);
+        assertEquals("1234@placeholder.email", email);
+    }
+
+    @Test
+    public void testGetPlaceholderEmailEnvironmentButNoIds() throws Exception {
+        EventProcessingRequest request = new EventProcessingRequest();
+        request.setRuntimeEnvironment(new AndroidRuntimeEnvironment());
+        request.setDeviceApplicationStamp("1234");
+        String email = IterableExtension.getPlaceholderEmail(request);
+        assertEquals("1234@placeholder.email", email);
+
+        request = new EventProcessingRequest();
+        request.setRuntimeEnvironment(new IosRuntimeEnvironment());
+        request.setDeviceApplicationStamp("12345");
+        email = IterableExtension.getPlaceholderEmail(request);
+        assertEquals("12345@placeholder.email", email);
+
+        request = new EventProcessingRequest();
+        request.setRuntimeEnvironment(new TVOSRuntimeEnvironment());
+        request.setDeviceApplicationStamp("123456");
+        email = IterableExtension.getPlaceholderEmail(request);
+        assertEquals("123456@placeholder.email", email);
+    }
+
```

Comment:
What about tvOS? I know that iOS and tvOS basically follow the same execution path but still might be good to have them. Up to you.